src/main/scala/chisel3/UnitTest.scala

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3.experimental.BaseModule
+import chisel3.experimental.hierarchy.Definition
+import java.io.File
+import java.util.jar.JarFile
+import scala.collection.JavaConverters._
+
+/** All classes and objects marked as `UnitTest` are automatically discoverable
+ *  by the `DiscoverUnitTests` helper. */
+trait UnitTest
+
+/** Helper to discover all subtypes of `UnitTest` in the class path, and call
+ *  their constructors (if they are a class) or ensure that the singleton is
+ *  constructed (if they are an object).
+ *
+ *  This code is loosely based on the test suite discovery in scalatest, which
+ *  performs the same scan over the classpath JAR files and directories, and
+ *  guesses class names based on the encountered directory structure. */
+object DiscoverUnitTests {
```

Comment:
Does this need to be public or can it be package private (`private[chisel3]`)?

Reply:
Either way works. No need for this to be public, except maybe some fringe users wanting to discover unit tests and perform their own action on them. But we could definitely keep this internal until someone asks for it.